dbcon/joblist/tuplehashjoin.cpp

```diff
@@ -200,140 +202,237 @@ void TupleHashJoinStep::join()
+        {
+            joinIsTooBig = true;
+            fLogger->logMessage(logging::LOG_TYPE_INFO, logging::ERR_JOIN_TOO_BIG);
+            errorMessage(logging::IDBErrorInfo::instance()->errorMsg(logging::ERR_JOIN_TOO_BIG));
+            status(logging::ERR_JOIN_TOO_BIG);
+            cout << "Join is too big, raise the UM join limit for now (monitor thread)" << endl;
+            abort();
+        }
+    }
+}
 
+void TupleHashJoinStep::startSmallRunners(uint index)
+{
+    utils::setThreadName("HJSStartSmall");
     string extendedInfo;
-    extendedInfo += toString();  // is each small side supposed to have the whole THJS info?
+    JoinType jt;
+    boost::shared_ptr<TupleJoiner> joiner;
 
-    smallDL = smallDLs[index];
-    smallIt = smallIts[index];
-    smallRG = smallRGs[index];
     jt = joinTypes[index];
+    extendedInfo += toString();
 
-    //cout << "    smallRunner " << index << " sees jointype " << jt << " joinTypes has " << joinTypes.size()
-    //	<< " elements" << endl;
     if (typelessJoin[index])
     {
-        joiner.reset(new TupleJoiner(smallRG, largeRG, smallSideKeys[index],
-                                     largeSideKeys[index], jt));
+        joiner.reset(new TupleJoiner(smallRGs[index], largeRG, smallSideKeys[index],
+                                     largeSideKeys[index], jt, &jobstepThreadPool));
     }
     else
     {
-        joiner.reset(new TupleJoiner(smallRG, largeRG, smallSideKeys[index][0],
-                                     largeSideKeys[index][0], jt));
+        joiner.reset(new TupleJoiner(smallRGs[index], largeRG, smallSideKeys[index][0],
+                                     largeSideKeys[index][0], jt, &jobstepThreadPool));
     }
 
     joiner->setUniqueLimit(uniqueLimit);
     joiner->setTableName(smallTableNames[index]);
     joiners[index] = joiner;
 
+    /* check for join types unsupported on the PM. */
+    if (!largeBPS || !isExeMgr)
+        joiner->setInUM(rgData[index]);
+
     /*
-    	read the small side into a TupleJoiner
-    	send the TupleJoiner to the large side TBPS
-    	start the large TBPS
-    	read the large side, write to the output
+        start the small runners
+        join them
+        check status
+        handle abort, out of memory, etc
     */
 
-    smallRG.initRow(&r);
-// 	cout << "reading smallDL" << endl;
-    more = smallDL->next(smallIt, &oneRG);
+    boost::posix_time::ptime end_time, start_time =
+        boost::posix_time::microsec_clock::universal_time();
+
+    stopMemTracking = false;
+    uint64_t jobs[numCores];
+    uint64_t memMonitor = jobstepThreadPool.invoke([this, index] { this->trackMem(index); });
+    // starting 1 thread when in PM mode, since it's only inserting into a
+    // vector of rows.  The rest will be started when converted to UM mode.
+    if (joiner->inUM())
+        for (int i = 0; i < numCores; i++)
+            jobs[i] = jobstepThreadPool.invoke([this, i, index, &jobs] { this->smallRunnerFcn(index, i, jobs); });
```

Comment:
Just to highlight. The first lambda in the code :)

Reply:
:man_dancing: